### PR TITLE
BentoML Blog Post

### PR DESCRIPTION
**Pull request by safoinme**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Briefly describe the changes you have introduced.

### CODE REVIEW

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
+---
+layout: post
+author: Safoine El khabich
+title: "Deploying Machine Learning Models from Local to Cloud with ZenML pipelines and BentoML"
+description: "Explore how to use ZenML pipelines and BentoML to easily deploy machine learning models from local to cloud environments. We will show how to train a model using ZenML, package it with BentoML, and deploy it to a local machine or cloud provider. By the end of this post, you will have a better understanding of how to streamline the deployment of your machine learning models using ZenML and BentoML."
+category: zenml
+tags: zenml bentoml
+publish_date: December 12th, 2022
+date: 2022-12-12T00:02:00Z
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:9]: same

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+thumbnail: /assets/posts/neptune_released/Release_0.23.0.gif
+image:
+  path: /assets/posts/neptune_released/Release_0.23.0.jpg
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:12]: for changing

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+To make this more easily understandable and cleaner we will be splitting the code into two different pipelines. The first one will be the training pipeline, which consists of the following six steps:
+
+1. **Load the Fashion-MNIST dataset**: This step loads the Fashion-MNIST dataset, which is a popular dataset for training image classification models.
+2. **Train a classifier**: In this step, we train a classifier on the Fashion-MNIST dataset using a machine learning algorithm of our choice. This could be a simple model like logistic regression or a more complex model like a convolutional neural network (CNN).
+3. **Evaluator**: After training the classifier, we use the Evaluator step to evaluate the performance of the model on a held-out test set. This allows us to assess the accuracy, precision, and other metrics of our model and helps us determine whether it is ready for deployment.
+4. **Deployment_trigger**: The deployment_trigger step is used to specify when and how the model should be deployed. This step allows us to set certain conditions, such as the performance of the model on the test set, that must be met before the model is deployed.
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:100]:
```suggestion
4. **Deployment_trigger**: The `deployment_trigger` step is used to specify when and how the model should be deployed. This step allows us to set certain conditions, such as the performance of the model on the test set, that must be met before the model is deployed.
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+5. **Bento_builder**: The bento_builder step takes the trained model and packages it into a format that can be easily deployed and served. This step is important because it allows us to easily share and deploy our trained model to different environments, such as a local machine or the cloud.
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:101]:
```suggestion
5. **Bento Builder**: The `bento_builder` step converts the trained model and packages it into a format that can be easily deployed and served. This step is important because it allows us to easily share and deploy our trained model to different environments, such as a local machine or the cloud.
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    """Link all the steps and artifacts together"""
+    train_dataloader, test_dataloader = importer()
+    model = trainer(train_dataloader)
+    accuracy = evaluator(test_dataloader=test_dataloader, model=model)
+    decision = deployment_trigger(accuracy=accuracy)
+    bento = bento_builder(model=model)
+    deployer(deploy_decision=decision, bento=bento)
+```
+
+then every step of these can be defined on its own we let’s take a look at how the trainer step for example is defined:
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:125]:
```suggestion
Then each of these steps can be defined on their own. Let’s take a look at how the trainer step is defined, for example:
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        loss.backward()
+        optimizer.step()
+
+        if batch % 100 == 0:
+            loss, current = loss.item(), batch * len(X)
+            print(f"loss: {loss:>7f}  [{current:>5d}/{size:>5d}]")
+    return model
+```
+
+Once the model is trained, the next step is the bento_builder step, which is a built-in ZenML step that makes it easier to prepare your model for deployment with BentoML. This step builds a Bento, which is an artifact that packages the model, requirements, and custom pre- and post-processing code in one file that can be used to deploy the model anywhere. The bento_builder step can be referenced as follows:
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:160]:
```suggestion
Once the model is trained, the next step is the `bento_builder` step, which is a built-in ZenML step that makes it easier to prepare your model for deployment with BentoML. This step builds a Bento, which is an artifact that packages the model, requirements, and custom pre- and post-processing code in one file that can be used to deploy the model anywhere. The `bento_builder` step can be referenced as follows:
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        port=3001,
+        production=False,
+    )
+)
+```
+After defining all these steps and the pipeline we can run the pipeline which will give us the following:
+
+![Pipeline terminal logs]({{ site.url }}/assets/posts/bentoml-deployment/pipeline-terminal-logs.png)
+
+In the ZenML dashboard we can see the pipeline dag:
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:208]:
```suggestion
In the ZenML dashboard we can see the pipeline DAG:
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+![Pipeline dag]({{ site.url }}/assets/posts/bentoml-deployment/pipeline-dag.png)
+
+In the Swagger UI, we can see more details about the deployed bento:
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:212]:
```suggestion
In the Swagger UI, we can see more details about the deployed Bento:
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+![Swagger UI]({{ site.url }}/assets/posts/bentoml-deployment/swagger-ui.png)
+
+## Inference Pipeline:
+
+To make use of our deployed model the example comes with an inference pipeline which is a set of steps that use some image samples to run prediction on it. You can find the code of this pipeline and more in the Bentoml_Deployment example.
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:218]:
```suggestion
To make use of our deployed model the example comes with an inference pipeline which is a set of steps that use some image samples to run predictions on. You can find the code of this pipeline and more in the [bentoml_deployment example](https://github.com/zenml-io/zenml/tree/main/examples/bentoml_deployment).
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+![Inference pipeline]({{ site.url }}/assets/posts/bentoml-deployment/inference-pipeline.png)
+
+## From Local to Cloud:
+
+Once a model server has been deployed and tested locally, it can be deployed to a cloud provider using Bentoctl, a command-line interface (CLI) tool for managing BentoML deployments on cloud providers. Bentoctl currently supports AWS, GCP, and Azure, and requires that you have an account with one of these providers as well as the corresponding CLI tools and Docker and Terraform installed on your machine.
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:224]:
```suggestion
Once a model server has been deployed and tested locally, it can be deployed to a cloud provider using `bentoctl`, a command-line interface (CLI) tool for managing BentoML deployments on cloud providers. Bentoctl currently supports AWS, GCP, and Azure, and requires that you have an account with one of these providers as well as the corresponding CLI tools and Docker and Terraform installed on your machine.
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+To use Bentoctl, you will need to first install it using the following command:
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:226]:
```suggestion
To use bentoctl, you will need to first install it using the following command:
```

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+```bash
+pip install bentoctl
+```
+
+Here is a corrected and improved version of the paragraph:
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:232]:
```suggestion
```

:)

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+
+Finally, we can deploy our model to AWS Sagemaker by running the apply command:
+
+```bash
+bentoctl apply
+```
+
+The full code for this example is available on [GitHub](https://github.com/zenml-io/zenml/tree/main/examples/bentoml_deployment).
+
+****Conclusion****
```
Comment (strickvl) [inline _posts/2022-12-12-bentoml-deployment-with-zenml.md:261]:
```suggestion
```

**Review by strickvl — CHANGES_REQUESTED** [review]
Nice work! Just some small changes which I think you can mostly just accept + merge them in.

File: _posts/2022-12-12-bentoml-deployment-with-zenml.md [hunk 14]
```diff
@@ -0,0 +1,269 @@
+---
+layout: post
+author: Safoine El khabich
+title: "Deploying Machine Learning Models from Local to Cloud with ZenML pipelines and BentoML"
```
[the remark this answers is not in the document]
Comment (safoinme) [reply _posts/2022-12-12-bentoml-deployment-with-zenml.md:4]: maybe "Local to Cloud Deployment with ZenML and BentoML"?

[the remark this answers is not in the document]
Comment (strickvl) [reply _posts/2022-12-12-bentoml-deployment-with-zenml.md]: Ok then either "Switching from Local ..." or just "From Local to Cloud...". Also welcome anyone else's opinion here. Maybe @htahir1 has some better ideas.

**Review by strickvl — APPROVED** [review]
LGTM!

**Review by htahir1 — APPROVED** [review]
